lib/email/receiver.rb

```diff
@@ -660,102 +660,97 @@ def trim_discourse_markers(reply)
       end
 
       return unless mail[:from]
 
       # For forwarded emails, where the from address matches a group incoming
       # email, we want to use the from address of the original email sender,
       # which we can extract from embedded_email_raw.
-      if has_been_forwarded?
-        if mail[:from].to_s =~ group_incoming_emails_regex && embedded_email[:from].errors.blank?
-          from_address, from_display_name = extract_from_fields_from_header(embedded_email, :from)
-          return from_address, from_display_name if from_address
+      if process_forwarded_emails && has_been_forwarded?
+        if mail[:from].to_s =~ group_incoming_emails_regex
+          if embedded_email && embedded_email[:from].errors.blank?
+            from_address, from_display_name =
+              Email::Receiver.extract_email_address_and_name(embedded_email[:from])
+            return from_address, from_display_name if from_address
+          end
         end
       end
 
+      # extract proper sender when using mailman mailing list
+      if mail[:x_mailman_version].present?
+        address, name = Email::Receiver.extract_email_address_and_name_from_mailman(mail)
+        return address, name if address
+      end
+
       # For now we are only using the Reply-To header if the email has
       # been forwarded via Google Groups, which is why we are checking the
       # X-Original-From header too. In future we may want to use the Reply-To
       # header in more cases.
-      if mail["X-Original-From"].present?
-        if mail[:reply_to] && mail[:reply_to].errors.blank?
-          from_address, from_display_name =
-            extract_from_fields_from_header(
-              mail,
-              :reply_to,
-              comparison_headers: ["X-Original-From"],
-            )
-          return from_address, from_display_name if from_address
-        end
-      end
-
-      if mail[:from].errors.blank?
-        from_address, from_display_name = extract_from_fields_from_header(mail, :from)
-        return from_address, from_display_name if from_address
+      if mail[:x_original_from].present? && mail[:reply_to].present?
+        original_from_address, _ =
+          Email::Receiver.extract_email_address_and_name(mail[:x_original_from])
+        reply_to_address, reply_to_name =
+          Email::Receiver.extract_email_address_and_name(mail[:reply_to])
+        return reply_to_address, reply_to_name if original_from_address == reply_to_address
       end
 
-      return extract_from_address_and_name(mail.from) if mail.from.is_a? String
-
-      if mail.from.is_a? Mail::AddressContainer
-        mail.from.each do |from|
-          from_address, from_display_name = extract_from_address_and_name(from)
-          return from_address, from_display_name if from_address
-        end
-      end
-
-      nil
+      Email::Receiver.extract_email_address_and_name(mail[:from])
```

Comment:
I have yet to see an email with more than one email address in `From`. If those happen in the wild, I'm happy to revert that part.